Tor support for Sharer - Sharer routes Getter requests through Tor

nirvantyagi opened the pull request:
This CL adds support for the Sharer to route Getter requests through Tor. It is done using two variables in advanced settings, tor boolean and tor port, that can be set to indicate to rtc-to-net to instead forward requests through a local tor socks proxy.

Things to still do:
- make code less tor specific -> more socks5 proxy general
- create integration test, write simple socks server
- create better UX support for tor settings/ advanced settings
- add username/password socks authentication for different Getters - Tor recommends different auth for different users (adds Tor specific logic)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2533)
<!-- Reviewable:end -->

bemasc commented:
Review status: 0 of 6 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/lib/integration-tests/socks-echo/proxy-integration-test.ts, line 18 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb1rIRkAcxFVT3v5EW:-KLb1rIRkAcxFVT3v5EX:-1940661146) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/integration-tests/socks-echo/proxy-integration-test.ts#L18)):*
> ```XML
> import ReceivedDataEvent = proxyintegrationtesttypes.ReceivedDataEvent;
> 
>   var log :logging.Log = new logging.Log('SocksClient');
> ```

Nit: should not be indented.

---

*[src/lib/rtc-to-net/proxyconfig.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb2HdBT4OrUlGu3DSR:-KLb2HdC9IhzSsx0TC5m:-496419579) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/proxyconfig.ts#L6)):*
> ```XML
>   // in a non-unicast (e.g. local network) address will fail.
>   allowNonUnicast :boolean;
>   torOn           :boolean;
> ```

As discussed, should be optional: `?:`

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 21 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLaqk_OVRrkb29vHwxR:-KLaqk_OVRrkb29vHwxS:-1268712121) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L21)):*
> ```XML
> 
>   var log :logging.Log = new logging.Log('RtcToNet');
>   var usingTor :boolean = true;
> ```

For name consistency, I would prefer the name to be `torOn`.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 385 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb2Nin2IRY4n3kkhJg:-KLb2Nin2IRY4n3kkhJh:193263680) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L385)):*
> ```XML
>     // close() methods should ever reject, this should never reject.
>     private stopResources_ = () : Promise<void> => {
>       log.info('%1: freeing resources', [this.longId()]);
> ```

Do you think this should be merged?

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 442 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb3807KiD49vW6YHD5:-KLb3807KiD49vW6YHD6:-1861255025) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L442)):*
> ```XML
> 
>     // Initiates tcp connection to endpoint
>     private getTcpConnection_ = (endpoint:net.Endpoint) :tcp.Connection => {
> ```

It seems that getTcpConnection, waitForTcpConnection, and handleTcpConnection all do radically different things depending on the value of `torOn`.  I think it would be more elegant to separate the tor and non-tor paths into their own methods, so that we only have to check the value of `torOn` in one place.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 450 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLaxTiZzLlVwpL6p-8-:-KLaxTiZzLlVwpL6p-80:-911299408) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L450)):*
> ```XML
>       }
>       this.webEndpoint_ = endpoint;
>       log.debug('%1: Creating tcp connection with tor settings: torOn: %2, torPort: %3', [this.longId(), this.proxyConfig_.torOn, this.proxyConfig_.torPort]);
> ```

Nit: Long line.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 457 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb2Tmw4oOQw0Z9e4r-:-KLb2TmxkCfQhlX_qkdQ:-1262950856) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L457)):*
> ```XML
>         var localTorEndpoint :net.Endpoint = {
>           address: '127.0.0.1',
>           port: this.proxyConfig_.torPort //default 9050
> ```

Nit: space after `//`

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 491 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb1aZoK27YgCQudRFt:-KLb1aZoK27YgCQudRFu:-2105726376) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L491)):*
> ```XML
>       if (this.proxyConfig_.torOn) {
>         log.debug('%1: Connected to local tor proxy', [this.longId()]);
>         //Send request through Socks protocol
> ```

Nit: Space after `//`

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 501 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb1fhHJnlEGfuRGm2U:-KLb1fhHJnlEGfuRGm2V:-380720168) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L501)):*
> ```XML
>         this.replyToPeer_(reply, info);
>       }
> ```

Nit: unneeded blank line

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 527 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb19OeSftcjB6C-Oqe:-KLb19OeSftcjB6C-Oqf:-1910120) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L527)):*
> ```XML
>           var request :socks.Request = {
>             command: socks.Command.TCP_CONNECT,
>             endpoint: this.webEndpoint_,
> ```

Nit: no trailing comma.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 537 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0zh58HJT-C4dWyUF:-KLb0zh58HJT-C4dWyUG:-434090165) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L537)):*
> ```XML
>           log.debug('%1: Received request response: %2',
>                     [this.longId(), response]);
>           if (response.reply != socks.Reply.SUCCEEDED) {
> ```

Nit: Prefer `===` and `!==`.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 552 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0in4KmL94z1kqyq3:-KLb0in5m8ZZvMHGKFsO:-1599405930) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L552)):*
> ```XML
>       var response :socks.Response = {
>         reply: reply,
>         endpoint: info ? info.bound : {"address":"0.0.0.0","port":0}
> ```

Nit: Prefer single quotes, and one space after `:` and `,`.  Try running `grunt tslint` to catch similar issues.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 623 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0dj0e51q8qOyRnGw:-KLb0dj0e51q8qOyRnGx:-1072624173) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L623)):*
> ```XML
>         this.bytesSentToPeer_.handle(data.byteLength);
>         this.channelSentBytes_ += data.byteLength;
> ```

Nit: accidental blank line?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

trevj commented:
Review status: 0 of 7 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[src/lib/integration-tests/socks-echo/proxy-integration-test.ts, line 120 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbDjrIzFirJ2D_OVJ6:-KLbDjrIzFirJ2D_OVJ7:-374026324) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/integration-tests/socks-echo/proxy-integration-test.ts#L120)):*
> ```XML
> 
>   private connectThroughSocks_ = (socksEndpoint:net.Endpoint, webEndpoint:net.Endpoint) : Promise<tcp.Connection> => {
>     log.info('%1: Connecting to the local tor proxy: %2', ["TEST", socksEndpoint]);
> ```

`TEST`?

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 449 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbIMl6a_ajP0Ge0wyo:-KLbIMl6a_ajP0Ge0wyp:-1038536178) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/rtc-to-net/rtc-to-net.ts#L449)):*
> ```XML
>                         endpoint.address);
>       }
>       this.webEndpoint_ = endpoint;
> ```

You don't need this private variable - in general, maintain as little state as possible.

You're tying yourself up in knots here, logic-wise - really, your new function `waitForTcpConnection_` is taking the place of `tcpConnection_.onceConnected` **which is no longer sufficient because you may need to "do the SOCKS thing" first**. Merge `handleTcpConnection_` with `waitForTcpConnection_` and have it do the SOCKS negotiation, if necessary. Then you won't have to repeat the call to `sendReply` (in fact, you'll barely have to change `start()` at all).

I think @bemasc is saying basically the same thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

fortuna commented on src/generic_core/globals.ts:
```diff
@@ -59,7 +59,9 @@ export var settings :uproxy_core_api.GlobalSettings = {
   shouldHijackDO: true,
-  crypto: true
+  crypto: true,
+  torOn: false,
```
I'd recommend putting all Tor config in a sub-object.

The reason why is because it makes it easier to propagate the information the Tor integration needs without having to pass around the entire global settings. The motivation is similar to [_Don’t include primitive types in a top-level request or response proto_](https://docs.google.com/document/d/1boTOKzkRwdr4I1uq2sa1ZAka1mzNadtGhSuuA_w5vSw/preview?hl=en&forcehl=1#heading=h.oyu8cui1gym4).

This will simplify a bunch of places, where you'll have one variable instead of two.
It also avoids some semantic issues, where a field (e.g. torPort) may not make sense if Tor is not enabled.

The settings could look like this when enabled:
````
  tor: { port: 9050 }
````

and like this when disabled:
````
  tor: null
````

You'll need to define a TorSettings class with the port number field.

nirvantyagi commented:
I wanted to get your feedback on two things before proceeding further.
1) Use of object containing all socks proxy settings
2) Separation of connection logic (socks proxy vs. direct)

---

Review status: 0 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[src/generic_core/globals.ts, line 63 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCETTu6Dcv1sHn7K-r2-63:-KM1YcSLOF9cU6h_ZCNL:-436222522) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/generic_core/globals.ts#L63)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> I'd recommend putting all Tor config in a sub-object.
> 
> 
> 
> The reason why is because it makes it easier to propagate the information the Tor integration needs without having to pass around the entire global settings. The motivation is similar to [_Don’t include primitive types in a top-level request or response proto_](https://docs.google.com/document/d/1boTOKzkRwdr4I1uq2sa1ZAka1mzNadtGhSuuA_w5vSw/preview?hl=en&forcehl=1#heading=h.oyu8cui1gym4).
> 
> 
> 
> This will simplify a bunch of places, where you'll have one variable instead of two.
> 
> It also avoids some semantic issues, where a field (e.g. torPort) may not make sense if Tor is not enabled.
> 
> 
> 
> The settings could look like this when enabled:
> 
> ````
> 
>   tor: { port: 9050 }
> 
> ````
> 
> 
> 
> and like this when disabled:
> 
> ````
> 
>   tor: null
> 
> ````
> 
> 
> 
> You'll need to define a TorSettings class with the port number field.
</details>
I've now added a socksProxySettings interface that is defined in uproxy_core_api.ts

Question @bemasc @trevj : This interface is also used in rtc-to-net/proxyconfig.ts, so I import uproxy_core_api there. Is that okay? This means that proxyconfig, which is in lib, is importing from generic_core. One alternative is to define to interface either in proxyconfig.ts or in a separate file in rtc-to-net, and import that into uproxy_core_api.ts (where the global settings are defined). Thoughts?

---

*[src/lib/integration-tests/socks-echo/proxy-integration-test.ts, line 18 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb1rIRkAcxFVT3v5EW:-KLbBOo6R7lOggLMzt5Q:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/integration-tests/socks-echo/proxy-integration-test.ts#L18)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: should not be indented.
</details>
Done.

---

*[src/lib/integration-tests/socks-echo/proxy-integration-test.ts, line 120 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbDjrIzFirJ2D_OVJ6:-KM1ZTp2yy5hivDYyyvV:610327451) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/integration-tests/socks-echo/proxy-integration-test.ts#L120)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> `TEST`?
</details>
Removed

---

*[src/lib/rtc-to-net/proxyconfig.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb2HdBT4OrUlGu3DSR:-KLbBSdwjTkNR6NMc0zA:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/proxyconfig.ts#L6)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> As discussed, should be optional: `?:`
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 21 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLaqk_OVRrkb29vHwxR:-KLbBbCAgmwv3Kr0-baP:1486146443) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L21)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> For name consistency, I would prefer the name to be `torOn`.
</details>
Removed since variable is not used - switched to using global.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 385 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb2Nin2IRY4n3kkhJg:-KLbC-OSLMSmXzE0cFJY:-1241653400) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L385)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Do you think this should be merged?
</details>
No, removed.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 442 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb3807KiD49vW6YHD5:-KLbC5PEcVjrzIDbMG1b:-1871456000) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L442)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> It seems that getTcpConnection, waitForTcpConnection, and handleTcpConnection all do radically different things depending on the value of `torOn`.  I think it would be more elegant to separate the tor and non-tor paths into their own methods, so that we only have to check the value of `torOn` in one place.
</details>
Agreed, see reply to Trevor's comment.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 450 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLaxTiZzLlVwpL6p-8-:-KLbCto2BL8Sn4vMvf3j:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L450)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: Long line.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 457 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb2Tmw4oOQw0Z9e4r-:-KLbDBCH5zAojUWkViSa:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L457)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: space after `//`
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 491 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb1aZoK27YgCQudRFt:-KLbDHOE2BWXy_55UpTm:-1414579577) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L491)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: Space after `//`
</details>
Done.:501

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 501 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb1fhHJnlEGfuRGm2U:-KLbDTkLW_HFe4E1Hp_7:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L501)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: unneeded blank line
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 527 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb19OeSftcjB6C-Oqe:-KLbEZbqfuIy6LEtIjV7:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L527)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: no trailing comma.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 537 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0zh58HJT-C4dWyUF:-KLbE_-EfKKBze8xd400:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L537)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: Prefer `===` and `!==`.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 552 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0in4KmL94z1kqyq3:-KLbF7-8IwqBvP29CQ1H:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L552)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: Prefer single quotes, and one space after `:` and `,`.  Try running `grunt tslint` to catch similar issues.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 623 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0dj0e51q8qOyRnGw:-KLbFEAZhJz6qEfhJN2j:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L623)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: accidental blank line?
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 449 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbIMl6a_ajP0Ge0wyo:-KM1ZWH_KDTK8M1GtXIh:-1434418265) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/rtc-to-net/rtc-to-net.ts#L449)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> You don't need this private variable - in general, maintain as little state as possible.
> 
> You're tying yourself up in knots here, logic-wise - really, your new function `waitForTcpConnection_` is taking the place of `tcpConnection_.onceConnected` **which is no longer sufficient because you may need to "do the SOCKS thing" first**. Merge `handleTcpConnection_` with `waitForTcpConnection_` and have it do the SOCKS negotiation, if necessary. Then you won't have to repeat the call to `sendReply` (in fact, you'll barely have to change `start()` at all).
> 
> I think @bemasc is saying basically the same thing.
</details>
@bemasc @trevj I've done a pretty big overhaul of this section. I tried to clean it up, separate the socksProxyConnection path and the directConnection path while also reusing common logic helper functions. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-:-KM1_HxorEVLplYfzkyF:560239255)*
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/generic_core/globals.ts, line 63 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCETTu6Dcv1sHn7K-r2-63:-KM4mlH2rneGNustghtP:-817727928) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/generic_core/globals.ts#L63)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> I've now added a socksProxySettings interface that is defined in uproxy_core_api.ts
> 
> Question @bemasc @trevj : This interface is also used in rtc-to-net/proxyconfig.ts, so I import uproxy_core_api there. Is that okay? This means that proxyconfig, which is in lib, is importing from generic_core. One alternative is to define to interface either in proxyconfig.ts or in a separate file in rtc-to-net, and import that into uproxy_core_api.ts (where the global settings are defined). Thoughts?
</details>
This seems like a reasonable structure to me.

`socksProxyOn` is a confusing name, because we are a socks proxy.  How about changing `socksProxySettings` to `reproxy`, and then changing `socksProxyOn` to `enabled`?

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 552 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0in4KmL94z1kqyq3:-KM4oe_7juCUvv3jYqx6:478245757) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L552)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Done.
</details>
Still missing spaces after `:` and `,`.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 449 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbIMl6a_ajP0Ge0wyo:-KM4pEH-NmMPe19hXzrp:507749654) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/rtc-to-net/rtc-to-net.ts#L449)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> @bemasc @trevj I've done a pretty big overhaul of this section. I tried to clean it up, separate the socksProxyConnection path and the directConnection path while also reusing common logic helper functions. 
</details>
I like it.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 350 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM4pxL2GMT0Ck8EUgQ8:-KM4pxL2GMT0Ck8EUgQ9:1232165794) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/lib/rtc-to-net/rtc-to-net.ts#L350)):*
> ```XML
>           // Returns socks reply and bound endpoint of connection
>           if (!this.isWebEndpointValid_(webEndpoint)) {
>             throw new Error('tried to connect to disallowed address: ' +
> ```

Does the `this.replyToPeer_(this.getReplyFromError_(e));` path still work?  Do the integration tests pass?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

fortuna commented:
Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/interfaces/uproxy_core_api.ts, line 84 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5XnIdxZxfke6Bbnzi:-KM5XnIdxZxfke6Bbnzj:-24905255) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L84)):*
> ```XML
> // Contains settings directing rtc-to-net server to go directly to net or
> // go through a socks proxy server (such as local Tor proxy).
> export interface SocksProxySettings {
> ```

Please comment each field. Interfaces should be well documented.

---

*[src/interfaces/uproxy_core_api.ts, line 85 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5XZFjkkBXwi9ylyhn:-KM5XZFkRZgZPYyVrQxh:1017231547) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L85)):*
> ```XML
> // go through a socks proxy server (such as local Tor proxy).
> export interface SocksProxySettings {
>   socksProxyOn  :boolean;
> ```

You can remove the `socksProxy` prefix, since you are already in the `SocksProxy` structure.

---

*[src/interfaces/uproxy_core_api.ts, line 87 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5Xv23MJtxrZ7Jh1i7:-KM5Xv23MJtxrZ7Jh1i8:-1494478834) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L87)):*
> ```XML
>   socksProxyOn  :boolean;
>   socksEndpoint :net.Endpoint;
>   isTor         :boolean; // using tor requires additional logic
> ```

For readability, please start comments with capital letters (that's the Google style guide)

---

*[src/interfaces/uproxy_core_api.ts, line 87 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5Y1SwjkJ4N98dhQMd:-KM5Y1SwjkJ4N98dhQMe:-2012366252) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L87)):*
> ```XML
>   socksProxyOn  :boolean;
>   socksEndpoint :net.Endpoint;
>   isTor         :boolean; // using tor requires additional logic
> ```

This comment can be improved. What does `isTor` mean? What exact thing "is Tor"?
We should consider make this an enum instead of a boolean.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

trevj commented:
Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 449 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbIMl6a_ajP0Ge0wyo:-KM5cleIE9aWBWxQDf8f:-619284883) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/rtc-to-net/rtc-to-net.ts#L449)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I like it.
</details>
Better.

But `getTcpConnection_` is strange: it's now a one-liner which wraps a call to a constructor in a `Promise.resolve`but neither caller needs it to be async! I don't think you needed to extract `isWebEndpointValid_`? (BTW, a confusing name: protocol is irrelevant to SOCKS).

And `waitForTcpConnection_` is unusual too: it sets `this.tcpConnection_` **and** sends an error to the SOCKS client, if one occurs - I think that logic was better-placed in `start` (higher up in the call stack). It also unconditionally logs "connection instantiated to local tor proxy" regardless of whether a SOCKS proxy is in use (and it may not be a Tor proxy!).

Sorry, wish I could sit down with you for a bit - basically, I'm trying to say I think you can get away with a couple of line changes to `start` + `connectWithSocksAuth_`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

fortuna commented on src/generic_core/globals.ts:
```diff
@@ -59,7 +59,8 @@ export var settings :uproxy_core_api.GlobalSettings = {
   shouldHijackDO: true,
-  crypto: true
+  crypto: true,
+  reproxy: {enabled: false, socksEndpoint: {address: '127.0.0.1', port: 9050}}
```
Minor nit: use line breaks to make this `reproxy` line more readable:
````
reproxy: {
  enabled: false,
  socksEndpoint: { address: '127.0.0.1', port: 9050 }
}
````

nirvantyagi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/generic_core/globals.ts, line 63 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCETTu6Dcv1sHn7K-r2-63:-KMa8H40Y2XdWKtGyKaR:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/generic_core/globals.ts#L63)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This seems like a reasonable structure to me.
> 
> `socksProxyOn` is a confusing name, because we are a socks proxy.  How about changing `socksProxySettings` to `reproxy`, and then changing `socksProxyOn` to `enabled`?
</details>
Done.

---

*[src/generic_core/globals.ts, line 63 \[r5\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCETTu6Dcv1sHn7K-r5-63:-KMjvK5-Rr5vyDp4JA2a:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/26e7b792c1d38d5e552c1a3a19dd0bea7f629534/src/generic_core/globals.ts#L63)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> Minor nit: use line breaks to make this `reproxy` line more readable:
> 
> ````
> 
> reproxy: {
> 
>   enabled: false,
> 
>   socksEndpoint: { address: '127.0.0.1', port: 9050 }
> 
> }
> 
> ````
</details>
Done.

---

*[src/interfaces/uproxy_core_api.ts, line 84 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5XnIdxZxfke6Bbnzi:-KMa5x1q_BZ19dDEbiTE:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L84)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> Please comment each field. Interfaces should be well documented.
</details>
Done.

---

*[src/interfaces/uproxy_core_api.ts, line 85 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5XZFjkkBXwi9ylyhn:-KMa5y0tXJkabFR6wReJ:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L85)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> You can remove the `socksProxy` prefix, since you are already in the `SocksProxy` structure.
</details>
Done.

---

*[src/interfaces/uproxy_core_api.ts, line 87 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5Xv23MJtxrZ7Jh1i7:-KMa63Oq3gYDRioPzFbF:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L87)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> For readability, please start comments with capital letters (that's the Google style guide)
</details>
Done.

---

*[src/interfaces/uproxy_core_api.ts, line 87 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM5Y1SwjkJ4N98dhQMd:-KMa6vu5KnZbacjcuZE7:291109308) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/interfaces/uproxy_core_api.ts#L87)):*
<details><summary><i>Previously, fortuna (Vinicius Fortuna) wrote…</i></summary>
> This comment can be improved. What does `isTor` mean? What exact thing "is Tor"?
> We should consider make this an enum instead of a boolean.
> 
</details>
Removed isTor field of interface. It is not being used in this pull request. I will add back in as enum for future pull request where Tor requires specific logic.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 552 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLb0in4KmL94z1kqyq3:-KMa88X825F9HnWOgusV:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/604e49a042c337a72170195d726c1b6749f030af/src/lib/rtc-to-net/rtc-to-net.ts#L552)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Still missing spaces after `:` and `,`.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 449 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLbIMl6a_ajP0Ge0wyo:-KMjveGFa2NTYzPdu95p:1637124556) ([raw file](https://github.com/uproxy/uproxy/blob/8c745c750ce3b104fd1c6a03e7a9d964d150e2a7/src/lib/rtc-to-net/rtc-to-net.ts#L449)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Better.
> 
> But `getTcpConnection_` is strange: it's now a one-liner which wraps a call to a constructor in a `Promise.resolve`but neither caller needs it to be async! I don't think you needed to extract `isWebEndpointValid_`? (BTW, a confusing name: protocol is irrelevant to SOCKS).
> 
> And `waitForTcpConnection_` is unusual too: it sets `this.tcpConnection_` **and** sends an error to the SOCKS client, if one occurs - I think that logic was better-placed in `start` (higher up in the call stack). It also unconditionally logs "connection instantiated to local tor proxy" regardless of whether a SOCKS proxy is in use (and it may not be a Tor proxy!).
> 
> Sorry, wish I could sit down with you for a bit - basically, I'm trying to say I think you can get away with a couple of line changes to `start` + `connectWithSocksAuth_`.
</details>
`start` now looks almost as it was originally before reproxy logic was added, as we discussed. This required pulling out the `isWebEndpointValid` fxn from `getTcpConnection`. It also requires two checks for whether reproxy is enabled. In this version, I have added the webEndpoint to state, since it is needed in the last part of the `start` promise chain. There doesn't seem to be a clean way of  being able to access it at that point without passing it down as an additional return value through the promise chain.

Let me know what you think!

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 350 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KM4pxL2GMT0Ck8EUgQ8:-KMa7SO49B2sTLu_lEXZ:838339301) ([raw file](https://github.com/uproxy/uproxy/blob/8147495d992d345f4826c2bf42554b93248b5755/src/lib/rtc-to-net/rtc-to-net.ts#L350)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Does the `this.replyToPeer_(this.getReplyFromError_(e));` path still work?  Do the integration tests pass?
</details>
Fixed, now integration tests pass.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

trevj commented on src/lib/rtc-to-net/rtc-to-net.ts:
```diff
@@ -344,24 +344,54 @@ import ProxyConfig = require('./proxyconfig');
+          // Connect to web endpoint directly or through socks proxy
+          if (typeof this.proxyConfig_.reproxy !== 'undefined' &&
+              this.proxyConfig_.reproxy.enabled) {
+            return Promise.resolve(this.getTcpConnection_(
```
Cool. You don't need the `Promise.resolve` - if you return a plain object, the next `then()` handler will receive it. So I think you could just drop `getTcpConnection_ ` altogether.

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.spec.ts, line 37 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk7Eo4IUAkGrL6jB4Z:-KMk7Eo5Sey4JH8-6Lc7:-430290182) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.spec.ts#L37)):*
> ```XML
> var mockRemoteEndpoint :net.Endpoint = {
>   // This address and port are both reserved for testing.
>   address: '192.0.2.111',
> ```

Alternatively, instead of faking out every call to isWebEndpointValid, we could just change this to an allowed endpoint like '8.8.8.8', or change the RtcToNet constructor arguments to disable address restriction.

---

*[src/lib/rtc-to-net/rtc-to-net.spec.ts, line 152 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk5GMdB5alSJYlaXtn:-KMk5GMe86aF_YgonBbR:328438008) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.spec.ts#L152)):*
> ```XML
>     spyOn(session, 'receiveEndpointFromPeer_').and.returnValue(Promise.resolve(mockRemoteEndpoint));
>     spyOn(session, 'replyToPeer_').and.returnValue(Promise.resolve());
>     spyOn(session, 'isWebEndpointValid_').and.returnValue(Promise.resolve(true));
> ```

Why are these promises?  `isWebEndpointValid_` returns synchronously.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 360 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk8AS7p6CL9FPqOM2F:-KMk8AS7p6CL9FPqOM2G:906890528) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L360)):*
> ```XML
> 
>           // Connect to web endpoint directly or through socks proxy
>           if (typeof this.proxyConfig_.reproxy !== 'undefined' &&
> ```

This will fail if `this.proxyConfig_.reproxy` is `null`.  Instead, I recommend `if (this.proxyConfig_.reproxy && this.proxyConfig_.reproxy.enabled) {`

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 662 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk4RYIAnrrgdAyJALp:-KMk4RYIAnrrgdAyJALq:328114206) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L662)):*
> ```XML
> 
>     // Checks validity of web endpoint of getter request
>     private isWebEndpointValid_ = (endpoint :net.Endpoint) :boolean => {
> ```

This method should be called something like `isWebEndpointAllowed_`.  It's checking for disallowed (i.e. localhost/local-net) IP addresses, not invalid ones.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 515 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMkDGfqbbf6FyOMuvHy:-KMkDGfrpZ6B3x_BUpGk:867755173) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L515)):*
> ```XML
>           log.debug('%1: Received auth handshake reply: %2',
>                     [this.longId(), auth]);
>           if (auth != socks.Auth.NOAUTH) {
> ```

Nit: `!==`

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 546 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMkDWjj7r5McBCBnkpP:-KMkDWjj7r5McBCBnkpQ:1733314323) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L546)):*
> ```XML
>       var response :socks.Response = {
>         reply: reply,
>         endpoint: bound ? bound : undefined
> ```

Nit: might prefer `bound || undefined`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

trevj commented on src/lib/rtc-to-net/rtc-to-net.ts:
```diff
@@ -344,24 +344,54 @@ import ProxyConfig = require('./proxyconfig');
+            return Promise.resolve(this.getTcpConnection_(
+                webEndpoint, true)); // start paused
+          }
+        }).then((connection :tcp.Connection) :Promise<tcp.ConnectionInfo> => {
```
I really think you can avoid returning tuples - but we can leave that for a followup.

nirvantyagi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.spec.ts, line 37 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk7Eo4IUAkGrL6jB4Z:-KMkSa-j6kVpFwqxdLVJ:-1061328160) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.spec.ts#L37)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Alternatively, instead of faking out every call to isWebEndpointValid, we could just change this to an allowed endpoint like '8.8.8.8', or change the RtcToNet constructor arguments to disable address restriction.
</details>
It looks like `192.0.2.11:1023` is reserved for the purposes of testing, so using a different endpoint like `8.8.8.8` might not be the right choice.

I could change the RtcToNet constructor to take on a new argument if you'd like. Thoughts?

---

*[src/lib/rtc-to-net/rtc-to-net.spec.ts, line 152 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk5GMdB5alSJYlaXtn:-KMkKWH9W5FG2FvapnV5:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.spec.ts#L152)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Why are these promises?  `isWebEndpointValid_` returns synchronously.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 360 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk8AS7p6CL9FPqOM2F:-KMkLIr4xyxwyRsdKNx4:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L360)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This will fail if `this.proxyConfig_.reproxy` is `null`.  Instead, I recommend `if (this.proxyConfig_.reproxy && this.proxyConfig_.reproxy.enabled) {`
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 362 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCEUAAVyT6Jxj_7z-r6-362:-KMkQUJ1eFHAUcn6oaol:415503150) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L362)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Cool. You don't need the `Promise.resolve` - if you return a plain object, the next `then()` handler will receive it. So I think you could just drop `getTcpConnection_ ` altogether.
</details>
Done. Promise wrapper is removed around `getTcpConnection`. `getTcpConnection_` was not dropped since it used to insert dummy connections in rtc-to-net spec. The Promise of the `then` statement is also not removed since we need to reject an error. Throwing an error would not be good, since we do not want to catch it. (Errors are currently being caught in the nested catch structure that we talked about - which was causing some problems).

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 368 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCEUAAVyT6Jxj_7z-r6-368:-KMkLM3iG2H78VIrq0Qw:199418154) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L368)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> I really think you can avoid returning tuples - but we can leave that for a followup.
</details>
Okay, I will ping you next week and we can chat about this.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 515 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMkDGfqbbf6FyOMuvHy:-KMkLz31tLiVdTMOvNOM:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L515)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: `!==`
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 546 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMkDWjj7r5McBCBnkpP:-KMkPiMPa3YO3T8PIXFV:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L546)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: might prefer `bound || undefined`.
</details>
Done.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 662 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk4RYIAnrrgdAyJALp:-KMkKlt71mhFsRGc2W3K:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L662)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This method should be called something like `isWebEndpointAllowed_`.  It's checking for disallowed (i.e. localhost/local-net) IP addresses, not invalid ones.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.spec.ts, line 37 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk7Eo4IUAkGrL6jB4Z:-KMkTigYoOjh1e6Wup2g:1635954070) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.spec.ts#L37)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> It looks like `192.0.2.11:1023` is reserved for the purposes of testing, so using a different endpoint like `8.8.8.8` might not be the right choice.
> 
> I could change the RtcToNet constructor to take on a new argument if you'd like. Thoughts?
</details>
There's already a constructor parameter for this: `proxyConfig.allowNonUnicast`.  if you set it to true you should be able to remove the new `spyOn` lines.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 362 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCEUAAVyT6Jxj_7z-r6-362:-KMk8WUz4isagogrrdvU:407379996) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L362)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Done. Promise wrapper is removed around `getTcpConnection`. `getTcpConnection_` was not dropped since it used to insert dummy connections in rtc-to-net spec. The Promise of the `then` statement is also not removed since we need to reject an error. Throwing an error would not be good, since we do not want to catch it. (Errors are currently being caught in the nested catch structure that we talked about - which was causing some problems).
</details>
I still see the `Promise.resolve()` call here.  Maybe your change hasn't made it into Github yet?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.spec.ts, line 37 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KMk7Eo4IUAkGrL6jB4Z:-KMk_mZCEFGryqZjG1Qf:1978399292) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.spec.ts#L37)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> There's already a constructor parameter for this: `proxyConfig.allowNonUnicast`.  if you set it to true you should be able to remove the new `spyOn` lines.
</details>
Done. Great, that did the trick.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 362 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCEUAAVyT6Jxj_7z-r6-362:-KMk_2aWYh6TzSWRxUJK:-1781868995) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L362)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I still see the `Promise.resolve()` call here.  Maybe your change hasn't made it into Github yet?
</details>
Ah, I ended up needing to keep the Promise.resolves since I need to keep the `then` return as a Promise. The reason for that is so that I can Promise.reject an error. I talked to Trevor a bit about this and the messy-ish error handling that the `start` method uses, and he prefers I don't try and clean it up at this time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 362 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCEUAAVyT6Jxj_7z-r6-362:-KMkaIFsXwDuSEQ9TAwN:74442596) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L362)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Ah, I ended up needing to keep the Promise.resolves since I need to keep the `then` return as a Promise. The reason for that is so that I can Promise.reject an error. I talked to Trevor a bit about this and the messy-ish error handling that the `start` method uses, and he prefers I don't try and clean it up at this time.
</details>
OK.  FYI, you can just `throw` the error, and the Promise system will automatically convert that into a reject.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 362 \[r6\]](https://reviewable.io:443/reviews/uproxy/uproxy/2533#-KLYLCEUAAVyT6Jxj_7z-r6-362:-KMkcGpSE-1rgQI9mKOx:-2116541197) ([raw file](https://github.com/uproxy/uproxy/blob/324f1a7d7b2a8d0e84e7c396be85f580647468e1/src/lib/rtc-to-net/rtc-to-net.ts#L362)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> OK.  FYI, you can just `throw` the error, and the Promise system will automatically convert that into a reject.
</details>
Done. Good to know! Just made the change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2533)*
<!-- Sent from Reviewable.io -->

trevj commented:
:+1: Thanks for bearing with us!
